bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetsNodeSyncImpl.java

```diff
@@ -0,0 +1,663 @@
+    /**
+     * @see RemoteAssetsNodeSync#syncAssetNodes()
+     */
+    @Override
+    public void syncAssetNodes() {
+        List<String> syncPaths = new ArrayList<>();
+        syncPaths.addAll(this.remoteAssetsConfig.getTagSyncPaths());
+        syncPaths.addAll(this.remoteAssetsConfig.getDamSyncPaths());
+
+        try {
+            for (String syncPath : syncPaths) {
+                this.session.refresh(true);
+                JSONObject topLevelJson = getJsonFromUri(syncPath);
+                Node topLevelSyncNode = getOrCreateNode(syncPath, (String) topLevelJson.get(JcrConstants.JCR_PRIMARYTYPE));
+                createOrUpdateNodes(topLevelJson, topLevelSyncNode);
+                this.session.save();
+            }
+        } catch (IOException e) {
+            LOG.error("IO Exception {}", e);
+        } catch (JSONException e) {
+            LOG.error("Json Exception {}", e);
+        } catch (RepositoryException e) {
+            LOG.error("Repository Exception {}", e);
+        }
+    }
+
+    /**
+     * Grab JSON from the provided path and sync to the JCR.
+     * @param nextPath String
+     * @param primaryType String
+     * @return Node
+     * @throws RepositoryException exception
+     */
+    private Node getOrCreateNode(final String nextPath, final String primaryType) throws RepositoryException {
+        Node node;
+
+        try {
+            node = JcrUtils.getNodeIfExists(nextPath, this.session);
+            if (node == null) {
+                node = JcrUtil.createPath(nextPath, primaryType, this.session);
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("New node '{}' created.", node.getPath());
+                }
+            } else {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Node '{}' retrieved from JCR.", node.getPath());
+                }
+            }
+        } catch (RepositoryException re) {
+            LOG.error("Repository Exception. Unable to get or create node '{}'", nextPath, re);
+            throw re;
+        }
+
+        return node;
+    }
+
+    /**
+     * Get {@link JSONObject} from URL response.
+     * @param path String
+     * @return JSONObject
+     * @throws IOException exception
+     * @throws JSONException exception
+     */
+    private JSONObject getJsonFromUri(final String path) throws IOException, JSONException {
+        try {
+            URI pathUi = new URI(null, null, path, null);
+            // we want to traverse the JCR one level at a time, hence the '1' selector.
+            URL url = new URL(this.remoteAssetsConfig.getServer() + pathUi.toString() + ".1.json");
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setRequestProperty("Authorization", String.format("Basic %s", RemoteAssets.encodeForBasicAuth(this.remoteAssetsConfig)));
+
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
```

Comment:
So this is a big change, but can you pull down XML and use Session.importXML(..) to import everything?
You could even include the binary data in the XML and not make the separate calls...

something like...
http://localhost:4502/content/dam/we-retail/en/activities/climbing/climber-gear-outdoor.jpg/jcr:content.xml

Reply:
See thread above...